docs/reference/advanced_configuration.rst
Advanced Configuration
======================

Full configuration options:

.. configuration-block::

    .. code-block:: yaml

        # Default configuration for extension with alias: "sonata_notification"
        sonata_notification:

            # Other backends you can use:
            #
            # sonata.notification.backend.postpone
            # sonata.notification.backend.doctrine
            # sonata.notification.backend.rabbitmq
            backend:              sonata.notification.backend.runtime

            # Example for using RabbitMQ
            #     - { queue: myQueue, recover: true, default: false, routing_key: the_routing_key, dead_letter_exchange: 'my.dead.letter.exchange' }
            #     - { queue: catchall, default: true }
            #
            # Example for using Doctrine
            #     - { queue: sonata_page, types: [sonata.page.create_snapshot, sonata.page.create_snapshots] }
            #     - { queue: catchall, default: true }
            queues:

                # The name of the queue
                queue:                ~ # Required

                # Set the name of the default queue
                default:              false

                # Only used by RabbitMQ
                #
                # Direct exchange with routing_key
                routing_key:          ''

                # Only used by RabbitMQ
                #
                # If set to true, the consumer will respond with a `basic.recover` when an exception occurs,
                # otherwise it will not respond at all and the message will be unacknowledged
                recover:              false

                # Only used by RabbitMQ
                #
                # If is set, failed messages will be rejected and sent to this exchange
                dead_letter_exchange:  null

                # Only used by Doctrine
                #
                # Defines types handled by the message backend
                types:                []
            backends:
                doctrine:
                    message_manager:      sonata.notification.manager.message.default

                    # The max age in seconds
                    max_age:              86400

                    # The delay in microseconds
                    pause:                500000

                    # The number of items on each iteration
                    batch_size:           10

                    # Raising errors level
                    states:
                        in_progress:          10
                        error:                20
                        open:                 100
                        done:                 10000
                rabbitmq:
                    exchange:             ~ # Required
                    connection:
                        host:                 localhost
                        port:                 5672
                        user:                 guest
                        pass:                 guest
                        vhost:                guest
                        console_url:          'http://localhost:55672/api'
                        factory_class:        \Enqueue\AmqpLib\AmqpConnectionFactory
            consumers:

                # If set to true, SwiftMailerConsumer and LoggerConsumer will be registered as services
                register_default:     true

            # Listeners attached to the IterateEvent
            # Iterate event is thrown on each command iteration
            #
            # Iteration listener class must implement Sonata\NotificationBundle\Event\IterationListener
            iteration_listeners: []
            class:
                message: App\Entity\SonataNotificationMessage
            admin:
                enabled: true
                message:
                    class: Sonata\NotificationBundle\Admin\MessageAdmin
                    controller: 'SonataNotificationBundle:MessageAdmin'
                    translation: SonataNotificationBundle

    .. code-block:: yaml

        # config/packages/doctrine.yaml

        doctrine:
            orm:
                entity_managers:
                    default:
                        mappings:
                            SonataNotificationBundle: ~

Changing AMQP transport
-----------------------

Sonata integrates with `queue interop`_ and by default uses `enqueue/amqp-lib`_.
Though you can pick your favorite library among:

* `enqueue/amqp-ext`_: use `Enqueue\AmqpExt\AmqpConnectionFactory` class.
* `enqueue/amqp-lib`_: use `Enqueue\AmqpLib\AmqpConnectionFactory` class.
* `enqueue/amqp-bunny`_: use `Enqueue\AmqpBunny\AmqpConnectionFactory` class.

For example if you decide to use `enqueue/amqp-bunny`_ transport,
run `composer require enqueue/amqp-ext:^0.8` and change `factory_class` option in the config:

.. configuration-block::

    .. code-block:: yaml

        # config/packages/sonata_notification.yaml

        sonata_notification:
            backends:
                rabbitmq:
                    connection:
                        factory_class: Enqueue\AmqpExt\AmqpConnectionFactory

.. _`queue interop`: https://github.com/queue-interop/queue-interop#amqp-interop
.. _`enqueue/amqp-lib`: https://github.com/php-enqueue/enqueue-dev/blob/master/docs/transport/amqp_lib.md
.. _`enqueue/amqp-ext`: https://github.com/php-enqueue/enqueue-dev/blob/master/docs/transport/amqp_ext.md
.. _`enqueue/amqp-bunny`: https://github.com/php-enqueue/enqueue-dev/blob/master/docs/transport/amqp_bunny.md



Resources/doc/reference/advanced_configuration.rst
Advanced Configuration
======================

Full configuration options:

.. configuration-block::

    .. code-block:: yaml

        # Default configuration for extension with alias: "sonata_notification"
        sonata_notification:

            # Other backends you can use:
            #
            # sonata.notification.backend.postpone
            # sonata.notification.backend.doctrine
            # sonata.notification.backend.rabbitmq
            backend:              sonata.notification.backend.runtime

            # Example for using RabbitMQ
            #     - { queue: myQueue, recover: true, default: false, routing_key: the_routing_key, dead_letter_exchange: 'my.dead.letter.exchange' }
            #     - { queue: catchall, default: true }
            #
            # Example for using Doctrine
            #     - { queue: sonata_page, types: [sonata.page.create_snapshot, sonata.page.create_snapshots] }
            #     - { queue: catchall, default: true }
            queues:

                # The name of the queue
                queue:                ~ # Required

                # Set the name of the default queue
                default:              false

                # Only used by RabbitMQ
                #
                # Direct exchange with routing_key
                routing_key:          ''

                # Only used by RabbitMQ
                #
                # If set to true, the consumer will respond with a `basic.recover` when an exception occurs,
                # otherwise it will not respond at all and the message will be unacknowledged
                recover:              false

                # Only used by RabbitMQ
                #
                # If is set, failed messages will be rejected and sent to this exchange
                dead_letter_exchange:  null

                # Only used by RabbitMQ
                #
                # If is set, failed messages will be routed to the queue using this key by dead-letter-exchange,
                # otherwise it will be requeued to the original queue if `dead-letter-exchange` is set
                #
                # If is set, the queue must be configured with this key as routing_key
                dead_letter_routing_key:  null

                # Only used by Doctrine
                #
                # Defines types handled by the message backend
                types:                []
            backends:
                doctrine:
                    message_manager:      sonata.notification.manager.message.default

                    # The max age in seconds
                    max_age:              86400

                    # The delay in microseconds
                    pause:                500000

                    # The number of items on each iteration
                    batch_size:           10

                    # Raising errors level
                    states:
                        in_progress:          10
                        error:                20
                        open:                 100
                        done:                 10000
                rabbitmq:
                    exchange:             ~ # Required
                    connection:
                        host:                 localhost
                        port:                 5672
                        user:                 guest
                        pass:                 guest
                        vhost:                guest
                        console_url:          'http://localhost:55672/api'
            consumers:

                # If set to true, SwiftMailerConsumer and LoggerConsumer will be registered as services
                register_default:     true

            # Listeners attached to the IterateEvent
            # Iterate event is thrown on each command iteration
            #
            # Iteration listener class must implement Sonata\NotificationBundle\Event\IterationListener
            iteration_listeners:  []
            class:
                message:              Application\Sonata\NotificationBundle\Entity\Message
            admin:
                enabled:              true
                message:
                    class:                Sonata\NotificationBundle\Admin\MessageAdmin
                    controller:           'SonataNotificationBundle:MessageAdmin'
                    translation:          SonataNotificationBundle

    .. code-block:: yaml

        doctrine:
            orm:
                entity_managers:
                    default:
                        mappings:
                            SonataNotificationBundle: ~
                            ApplicationSonataNotificationBundle: ~


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [3.9.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.8.0...3.9.0) - 2020-07-26
### Added
- [[#448](https://github.com/sonata-project/SonataNotificationBundle/pull/448)]
  Added public alias
`Sonata\NotificationBundle\Controller\Api\MessageController` for
`sonata.notification.controller.api.message` service
([@wbloszyk](https://github.com/wbloszyk))

### Changed
- [[#452](https://github.com/sonata-project/SonataNotificationBundle/pull/452)]
  SonataEasyExtendsBundle is now optional, using SonataDoctrineBundle is
preferred ([@jordisala1991](https://github.com/jordisala1991))
- [[#452](https://github.com/sonata-project/SonataNotificationBundle/pull/452)]
  Use Laminas instead of deprecated Zend
([@jordisala1991](https://github.com/jordisala1991))

### Deprecated
- [[#452](https://github.com/sonata-project/SonataNotificationBundle/pull/452)]
  Using SonataEasyExtendsBundle to add Doctrine mapping information
([@jordisala1991](https://github.com/jordisala1991))

### Fixed
- [[#448](https://github.com/sonata-project/SonataNotificationBundle/pull/448)]
  Fix RestFul API - `Class could not be determined for Controller identified`
Error ([@wbloszyk](https://github.com/wbloszyk))

### Removed
- [[#451](https://github.com/sonata-project/SonataNotificationBundle/pull/451)]
  Removed support for deprecated "rest" routing type
([@wbloszyk](https://github.com/wbloszyk))
- [[#450](https://github.com/sonata-project/SonataNotificationBundle/pull/450)]
  Support for PHP < 7.2 ([@wbloszyk](https://github.com/wbloszyk))

## [3.8.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.7.0...3.8.0) - 2020-06-26
### Removed
- [[#440](https://github.com/sonata-project/SonataNotificationBundle/pull/440)]
  Remove SonataCoreBundle dependencies
([@wbloszyk](https://github.com/wbloszyk))

## [3.7.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.6.2...3.7.0) - 2020-03-26
### Fixed
- Fix doctrine deprecations

### Removed
- SonataEasyExtendsBundle
- Support for Symfony < 3.4
- Support for Symfony >= 4, < 4.2

## [3.6.2](https://github.com/sonata-project/SonataNotificationBundle/compare/3.6.1...3.6.2) - 2019-07-20
### Changed
- Make `sonata.notification.consumer.metadata` service public

## [3.6.1](https://github.com/sonata-project/SonataNotificationBundle/compare/3.6.0...3.6.1) - 2019-03-13
### Fixed
`TypeError` when `AMQPBackend`'s `publish` method is called.

## [3.6.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.5.1...3.6.0) - 2018-05-25

### Added
- Added `Sonata\NotificationBundle\Backend\BackendInterface` service alias
- Added `Sonata\NotificationBundle\Entity\MessageManager` service alias

### Changed
- Update RestartCommand check count message only for not pulling mode
- sendEmail function to allow for the setting of a return path
- Do not use deprecated `AMQPBackendDispatcher::getChannel` method.
- made the `sonata.notification.dispatcher` service public to fix a bug when running `sonata:notification:start`

### Fixed
- Fix deprecation for symfony/config 4.2+
- Make services public

### Removed
- support for php 5 and php 7.0

## [3.5.1](https://github.com/sonata-project/SonataNotificationBundle/compare/3.5.0...3.5.1) - 2018-05-25
# Changed
- Force use existing translation strings in breadcrumb for Message entity in Admin panel
- `enqueue/amqp-lib` is an optional dependency now

# Fixed
- API and Admin services are only available when using Doctrine as a backend

## [3.5.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.4.0...3.5.0) - 2018-04-26
### Added
- Added possibility to add an attachment to SwiftMailer Consumer

### Fixed
- Data fetched from stats counts are now properly manipulated (in case of doctrine backend is used)
- Typo in message status

## [3.4.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.3.1...3.4.0) - 2018-02-23
### Changed
- Require symfony/security-core instead of symfony/security
- Refactored bundle configuration
- Notification backend services are marked as public

### Fixed
- `each()` is deprecated since PHP 7.2
- Remove var **definition** override
- Commands not working on symfony4

### Removed
- Removed compatibility with older versions of FOSRestBundle (<2.1)

## [3.3.1](https://github.com/sonata-project/SonataNotificationBundle/compare/3.3.0...3.3.1) - 2018-01-26
### Changed
- Auto-register all aliases as public
- Auto-register consumer as public service
 
### Fixed
- `isRequired()` was removed since a default is specified
- MessageAdmin loads correct ChoiceType for the state filter (instead of ChoiceFilter)

## [3.3.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.2.0...3.3.0) - 2017-12-07
### Added
- Added Russian translations

### Changed
- Migrate from php-amqplib/php-amqplib to enqueue/amqp-lib package

### Fixed
- It is now allowed to install Symfony 4

### Removed
- Support for old versions of PHP and Symfony.
- Support deprecations for old form alias usage

## [3.2.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.1.0...3.2.0) - 2017-09-14
### Added
- Support for CC/BCC fields in `SwiftMailerConsumer`
- Added prefetch count configuration in AMQPBackend

### Fixed
- Fixed hardcoded paths to classes in `.xml.skeleton` files of config
- Use `EventDispatcher` instead of deprecated `ContainerAwareEventDispatcher`

## [3.1.0](https://github.com/sonata-project/SonataNotificationBundle/compare/3.0.0...3.1.0) - 2017-02-03
### Added
- Add dead letter handling in AMQPBackend
- Added per-queue message TTL in AMQPBackend

### Changed
- Changes the name of the vendor videlalvaro/php-amqplib to its new name php-amqplib/php-amqplib
- dependency from `guzzle/guzzle` to `guzzlehttp/guzzle`, because it is deprecated
- array `QueryParam` parameter to map

src/DependencyInjection/Configuration.php
<?php

declare(strict_types=1);

/*
 * This file is part of the Sonata Project package.
 *
 * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Sonata\NotificationBundle\DependencyInjection;

use Enqueue\AmqpLib\AmqpConnectionFactory;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\NodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder('sonata_notification');

        $rootNode = $treeBuilder->getRootNode()->children();

        $backendInfo = <<<'EOF'
Other backends you can use:

sonata.notification.backend.postpone
sonata.notification.backend.doctrine
sonata.notification.backend.rabbitmq
EOF;

        $iterationListenersInfo = <<<EOF
Listeners attached to the IterateEvent
Iterate event is thrown on each command iteration

Iteration listener class must implement Sonata\NotificationBundle\Event\IterationListener
EOF;

        $rootNode
            ->scalarNode('backend')
                ->info($backendInfo)
                ->defaultValue('sonata.notification.backend.runtime')
            ->end()
            ->append($this->getQueueNode())
            ->arrayNode('backends')
                ->children()
                    ->arrayNode('doctrine')
                        ->children()
                            ->scalarNode('message_manager')
                                ->defaultValue('sonata.notification.manager.message.default')
                            ->end()
                            ->scalarNode('max_age')
                                ->info('The max age in seconds')
                                ->defaultValue(86400)
                            ->end()
                            ->scalarNode('pause')
                                ->info('The delay in microseconds')
                                ->defaultValue(500000)
                            ->end()
                            ->scalarNode('batch_size')
                                ->info('The number of items on each iteration')
                                ->defaultValue(10)
                            ->end()
                            ->arrayNode('states')
                                ->info('Raising errors level')
                                ->addDefaultsIfNotSet()
                                ->children()
                                    ->integerNode('in_progress')
                                        ->defaultValue(10)
                                    ->end()
                                    ->integerNode('error')
                                        ->defaultValue(20)
                                    ->end()
                                    ->integerNode('open')
                                        ->defaultValue(100)
                                    ->end()
                                    ->integerNode('done')
                                        ->defaultValue(10000)
                                    ->end()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                    ->arrayNode('rabbitmq')
                        ->children()
                            ->scalarNode('exchange')
                                ->cannotBeEmpty()
                                ->isRequired()
                            ->end()
                            ->arrayNode('connection')
                                ->addDefaultsIfNotSet()
                                ->children()
                                    ->scalarNode('host')
                                        ->defaultValue('localhost')
                                    ->end()
                                    ->scalarNode('port')
                                        ->defaultValue(5672)
                                    ->end()
                                    ->scalarNode('user')
                                        ->defaultValue('guest')
                                    ->end()
                                    ->scalarNode('pass')
                                        ->defaultValue('guest')
                                    ->end()
                                    ->scalarNode('vhost')
                                        ->defaultValue('guest')
                                    ->end()
                                    ->scalarNode('console_url')
                                        ->defaultValue('http://localhost:55672/api')
                                    ->end()
                                    ->scalarNode('factory_class')
                                        ->cannotBeEmpty()
                                        ->defaultValue(AmqpConnectionFactory::class)
                                        ->info('This option defines an AMQP connection factory to be used to establish a connection with RabbitMQ.')
                                    ->end()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
            ->arrayNode('consumers')
                ->addDefaultsIfNotSet()
                ->children()
                    ->booleanNode('register_default')
                        ->info('If set to true, SwiftMailerConsumer and LoggerConsumer will be registered as services')
                        ->defaultTrue()
                    ->end()
                ->end()
            ->end()
            ->arrayNode('iteration_listeners')
                ->info($iterationListenersInfo)
                ->defaultValue([])
                ->prototype('scalar')->end()
            ->end()
            ->arrayNode('class')
                ->addDefaultsIfNotSet()
                ->children()
                    ->scalarNode('message')
                        ->defaultValue('App\\Entity\\Message')
                    ->end()
                ->end()
            ->end()
            ->arrayNode('admin')
                ->addDefaultsIfNotSet()
                ->children()
                    ->booleanNode('enabled')
                        ->defaultFalse()
                    ->end()
                    ->arrayNode('message')
                        ->addDefaultsIfNotSet()
                        ->children()
                            ->scalarNode('class')
                                ->cannotBeEmpty()
                                ->defaultValue('Sonata\\NotificationBundle\\Admin\\MessageAdmin')
                            ->end()
                            ->scalarNode('controller')
                                ->cannotBeEmpty()
                                ->defaultValue('SonataNotificationBundle:MessageAdmin')
                            ->end()
                            ->scalarNode('translation')
                                ->cannotBeEmpty()
                                ->defaultValue('SonataNotificationBundle')
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;

        return $treeBuilder;
    }

    /**
     * @return ArrayNodeDefinition|NodeDefinition
     */
    protected function getQueueNode()
    {
        $treeBuilder = new TreeBuilder('queues');

        $node = $treeBuilder->getRootNode();

        $queuesInfo = <<<'EOF'
Example for using RabbitMQ
    - { queue: myQueue, recover: true, default: false, routing_key: the_routing_key, dead_letter_exchange: 'my.dead.letter.exchange' }
    - { queue: catchall, default: true }

Example for using Doctrine
    - { queue: sonata_page, types: [sonata.page.create_snapshot, sonata.page.create_snapshots] }
    - { queue: catchall, default: true }
EOF;

        $routingKeyInfo = <<<'EOF'
Only used by RabbitMQ

Direct exchange with routing_key
EOF;

        $recoverInfo = <<<'EOF'
Only used by RabbitMQ

If set to true, the consumer will respond with a `basic.recover` when an exception occurs,
otherwise it will not respond at all and the message will be unacknowledged
EOF;

        $deadLetterExchangeInfo = <<<'EOF'
Only used by RabbitMQ

If is set, failed messages will be rejected and sent to this exchange
EOF;

        $deadLetterRoutingKeyInfo = <<<'EOF'
Only used by RabbitMQ

If set, failed messages will be routed to the queue using this key by dead-letter-exchange,
otherwise it will be requeued to the original queue if `dead-letter-exchange` is set.

If set, the queue must be configured with this key as `routing_key`.
EOF;

        $ttlInfo = <<<'EOF'
Only used by RabbitMQ

Defines the per-queue message time-to-live (milliseconds)
EOF;

        $prefetchCountInfo = <<<'EOF'
Only used by RabbitMQ

Defines the number of messages which will be delivered to the customer at a time.
EOF;

        $typesInfo = <<<'EOF'
Only used by Doctrine

Defines types handled by the message backend
EOF;

        $connectionNode = $node
            ->info($queuesInfo)
            ->requiresAtLeastOneElement()
            ->prototype('array')
        ;

        $connectionNode
            ->children()
                ->scalarNode('queue')
                    ->info('The name of the queue')
                    ->cannotBeEmpty()
                    ->isRequired()
                ->end()
                ->booleanNode('default')
                    ->info('Set the name of the default queue')
                    ->defaultValue(false)
                ->end()

                // RabbitMQ configuration
                ->scalarNode('routing_key')
                    ->info($routingKeyInfo)
                    ->defaultValue('')
                ->end()
                ->booleanNode('recover')
                    ->info($recoverInfo)
                    ->defaultValue(false)
                ->end()
                ->scalarNode('dead_letter_exchange')
                    ->info($deadLetterExchangeInfo)
                    ->defaultValue(null)
                ->end()
                ->scalarNode('dead_letter_routing_key')
                    ->info($deadLetterRoutingKeyInfo)
                    ->defaultValue(null)
                ->end()
                ->integerNode('ttl')
                    ->info($ttlInfo)
                    ->min(0)
                    ->defaultValue(null)
                ->end()
                ->integerNode('prefetch_count')
                    ->info($prefetchCountInfo)
                    ->min(0)
                    ->max(65535)
                    ->defaultValue(null)
                ->end()

                // Database configuration (Doctrine)
                ->arrayNode('types')
                    ->info($typesInfo)
                    ->defaultValue([])
                    ->prototype('scalar')->end()
                ->end()
            ->end();

        return $node;
    }
}
